theta_dup/include/theta_sketch_dup.h

```diff
@@ -0,0 +1,2209 @@
+   * sketch
+   * @return an instance of a sketch as a unique_ptr
+   */
+  static unique_ptr deserialize(std::istream& is, uint64_t seed = DEFAULT_SEED);
+
+  /**
+   * This method deserializes a sketch from a protobuf.
+   * @param pb input stream
+   * @param seed the seed for the hash function that was used to create the
+   * sketch
+   * @return an instance of a sketch as a unique_ptr
+   */
+  static unique_ptr deserialize(datasketches_pb::Update_theta_sketch_dup& pb,
+                                uint64_t seed = DEFAULT_SEED);
+
+  /**
+   * This method deserializes a sketch from a given array of bytes.
+   * @param bytes pointer to the array of bytes
+   * @param size the size of the array
+   * @param seed the seed for the hash function that was used to create the
+   * sketch
+   * @return an instance of the sketch
+   */
+  static unique_ptr deserialize(const void* bytes, size_t size,
+                                uint64_t seed = DEFAULT_SEED);
+
+  class const_iterator;
+
+  /**
+   * Iterator over hash values in this sketch.
+   * @return begin iterator
+   */
+  virtual const_iterator begin() const = 0;
+
+  /**
+   * Iterator pointing past the valid range.
+   * Not to be incremented or dereferenced.
+   * @return end iterator
+   */
+  virtual const_iterator end() const = 0;
+
+  /**
+   * @return true if *this equals r
+   */
+  virtual bool isEqual(const theta_sketch_dup_alloc& r) const;
+
+ protected:
+  enum flags { IS_BIG_ENDIAN, IS_READ_ONLY, IS_EMPTY, IS_COMPACT, IS_ORDERED };
+
+  bool is_empty_;
+  uint64_t theta_;
+
+  theta_sketch_dup_alloc(bool is_empty, uint64_t theta);
+
+  static uint16_t get_seed_hash(uint64_t seed);
+
+  static void check_sketch_type(uint8_t actual, uint8_t expected);
+  static void check_serial_version(uint8_t actual, uint8_t expected);
+  static void check_seed_hash(uint16_t actual, uint16_t expected);
+
+  /* TODO: support set operations
+   * friend theta_intersection_alloc<A>;
+   * friend theta_a_not_b_alloc<A>;
+   */
+};
+
+// update sketch
+
+template <typename A>
+using AllocU64 = typename std::allocator_traits<A>::template rebind_alloc<
+    std::pair<uint64_t, uint64_t>>;
+template <typename A>
+using vector_u64 = std::vector<std::pair<uint64_t, uint64_t>, AllocU64<A>>;
+
+template <typename A>
+class update_theta_sketch_dup_alloc : public theta_sketch_dup_alloc<A> {
+ public:
+  class builder;
+  enum resize_factor { X1, X2, X4, X8 };
+  static const uint8_t SKETCH_TYPE = 2;
+
+  // No constructor here. Use builder instead.
+
+  virtual ~update_theta_sketch_dup_alloc() = default;
+
+  virtual uint32_t get_num_retained() const;
+  virtual uint16_t get_seed_hash() const;
+  virtual bool is_ordered() const;
+  virtual string<A> to_string(bool print_items = false) const;
+  // serialize to output stream
+  virtual void serialize(std::ostream& os) const;
+  // serialize to protobuf
+  virtual void serialize(datasketches_pb::Update_theta_sketch_dup* pb) const;
```

Comment:
serialization of the **derived class** update_theta_sketch_dup_alloc.
dp: the variable of the protobuf generated class.